lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/TxWithUTxO.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE ExplicitNamespaces #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Copyright: © 2024 Cardano Foundation
+-- License: Apache-2.0
+--
+-- Provides the 'TxWithUTxO' data type.
+--
+module Internal.Cardano.Write.Tx.TxWithUTxO
+    ( type TxWithUTxO
+    , pattern TxWithUTxO
+    , isValid
+    )
+    where
+
+import Prelude
+
+import Cardano.Ledger.Api
+    ( AlonzoEraTxBody (collateralInputsTxBodyL)
+    , EraTx (bodyTxL)
+    , EraTxBody (TxBody, inputsTxBodyL)
+    )
+import Control.Lens
+    ( over
+    , view
+    )
+import Data.Set
+    ( Set
+    )
+import Internal.Cardano.Write.Tx
+    ( IsRecentEra
+    , Tx
+    , TxIn
+    , UTxO (UTxO)
+    )
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+
+-- | A transaction with an associated UTxO set.
+--
+-- Every input in the transaction is guaranteed to resolve to a UTxO within the
+-- associated UTxO set.
+--
+-- The UTxO set may also contain additional UTxOs that are not referenced by
+-- the transaction.
+--
+data TxWithUTxO era = UnsafeTxWithUTxO (Tx era) (UTxO era)
+
+{-# COMPLETE TxWithUTxO #-}
+pattern TxWithUTxO :: IsRecentEra era => Tx era -> UTxO era -> TxWithUTxO era
+pattern
+    TxWithUTxO tx utxo <- UnsafeTxWithUTxO tx utxo where
+    TxWithUTxO tx utxo = construct tx utxo
+
+-- | Constructs a 'TxWithUTxO' object.
+--
+-- The given transaction will be stripped of any inputs that do not resolve to
+-- UTxOs with the given UTxO set.
+--
+construct
+    :: forall era. IsRecentEra era
+    => Tx era
+    -> UTxO era
+    -> TxWithUTxO era
+construct tx0 utxo@(UTxO utxoMap) = UnsafeTxWithUTxO tx utxo
+  where
+    tx :: Tx era
+    tx = over bodyTxL removeUnresolvableInputs tx0
+
+    removeUnresolvableInputs :: TxBody era -> TxBody era
+    removeUnresolvableInputs
+        = over           inputsTxBodyL f
+        . over collateralInputsTxBodyL f
```

Comment:
> Notably missing: referenceInputsTxBodyL 😄

Thanks! I've now added this.

> It would be nice to use allInputsTxBodyF

Good point. I've used `allInputsTxBodyF` in the definition of `unresolvedInputs`.